File: Packages/StackEngineCore/Tests/StackEngineCoreTests/ImagePyramidTests.swift
import XCTest
import simd
@testable import StackEngineCore

final class ImagePyramidTests: XCTestCase {
    func testReduceHalvesDimensions(){
        let img = PixelImage(width: 8, height: 6, fill: SIMD3<Float>(0.5, 0.5, 0.5))
        let r = ImagePyramid.reduce(img)
        XCTAssertEqual(r.width, 4); XCTAssertEqual(r.height, 3)   // ceil(w/2), ceil(h/2)
    }

    func testReduceOfConstantIsConstant() {
        let img = PixelImage(width: 8, height: 8, fill: SIMD3<Float>(0.3, 0.6, 0.9))
        let r = ImagePyramid.reduce(img)
        XCTAssertEqual(r[1, 1].x, 0.3, accuracy: 1e-5)   // border-renormalized → no darkening
        XCTAssertEqual(r[1, 1].y, 0.6, accuracy: 1e-5)
        XCTAssertEqual(r[0, 0].z, 0.9, accuracy: 1e-5)   // even at the corner
    }

    func testExpandOfConstantIsConstantAtTargetSize() {
        let img = PixelImage(width: 4, height: 4, fill: SIMD3<Float>(0.4, 0.4, 0.4))
        let e = ImagePyramid.expand(img, toWidth: 8, toHeight: 7)
        XCTAssertEqual(e.width, 8); XCTAssertEqual(e.height, 7)
        XCTAssertEqual(e[3, 3].x, 0.4, accuracy: 1e-5)
    }

    func testLaplacianCollapseReconstructsConstant() {
        // For a constant image, reduce/expand are exact → collapse(laplacian(img)) == img exactly.
        let img = PixelImage(width: 12, height: 10, fill: SIMD3<Float>(0.25, 0.5, 0.75))
        let out = ImagePyramid.collapse(ImagePyramid.laplacian(img))
        XCTAssertEqual(out.width, 12); XCTAssertEqual(out.height, 10)
        XCTAssertLessThan(Metrics.maxAbsDiff(out, img), 1e-4)
    }
}

File: android/stackengine/src/test/kotlin/com/jdpneto/stackengine/ImagePyramidTests.kt
package com.jdpneto.stackengine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ImagePyramidTests {

    @Test
    fun testReduceHalvesDimensions() {
        val img = PixelImage(8, 6, Vec3(0.5f, 0.5f, 0.5f))
        val r = ImagePyramid.reduce(img)
        assertEquals(4, r.width)   // ceil(w/2)
        assertEquals(3, r.height)  // ceil(h/2)
    }

    @Test
    fun testReduceOfConstantIsConstant() {
        val img = PixelImage(8, 8, Vec3(0.3f, 0.6f, 0.9f))
        val r = ImagePyramid.reduce(img)
        assertEquals(0.3f, r[1, 1].x, absoluteTolerance = 1e-5f)   // border-renormalized → no darkening
        assertEquals(0.6f, r[1, 1].y, absoluteTolerance = 1e-5f)
        assertEquals(0.9f, r[0, 0].z, absoluteTolerance = 1e-5f)   // even at the corner
    }

    @Test
    fun testExpandOfConstantIsConstantAtTargetSize() {
        val img = PixelImage(4, 4, Vec3(0.4f, 0.4f, 0.4f))
        val e = ImagePyramid.expand(img, 8, 7)
        assertEquals(8, e.width)
        assertEquals(7, e.height)
        assertEquals(0.4f, e[3, 3].x, absoluteTolerance = 1e-5f)
    }

    @Test
    fun testLaplacianCollapseReconstructsConstant() {
        // For a constant image, reduce/expand are exact → collapse(laplacian(img)) == img exactly.
        val img = PixelImage(12, 10, Vec3(0.25f, 0.5f, 0.75f))
        val out = ImagePyramid.collapse(ImagePyramid.laplacian(img))
        assertEquals(12, out.width)
        assertEquals(10, out.height)
        assertTrue(Metrics.maxAbsDiff(out, img) < 1e-4f)
    }
}
